luna/gateware/platform/numato_opsis.py

```diff
@@ -0,0 +1,208 @@
+        )
+
+        m.submodules.feedback_bufg =  Instance("BUFG",
+            i_I = ClockSignal("feedback_out"),
+            o_O = ClockSignal("feedback")
+        )
+
+        m.submodules.usb_bufg =  Instance("BUFG",
+            i_I = ClockSignal("usb_pll_local"),
+            o_O = ClockSignal("usb")
+        )
+
+        m.submodules.usb_shifted_bufg =  Instance("BUFG",
+            i_I = ClockSignal("usb_pll_local_shifted"),
+            o_O = ClockSignal("usb_shifted")
+        )
+
+        # bring up ULPI clock
+        #debug_clk = platform.request("debug_header", 2, dir='o', xdr=2)
+
+        m.d.comb += [
+            #debug_clk.o_clk.eq(ClockSignal("usb")),
+            #debug_clk.o0.eq(0),
+            #debug_clk.o1.eq(1),
+            forward_ulpi_clk.o_clk.eq(ClockSignal("usb_shifted")),
+            forward_ulpi_clk.o0.eq(0),
+            forward_ulpi_clk.o1.eq(1),
+            ClockSignal("sync")  .eq(ClockSignal("usb")),
+            ClockSignal("fast")  .eq(platform.request("clk_100MHz")),
+            ResetSignal("usb")   .eq(~usb2_locked),
+            ResetSignal("sync")  .eq(~usb2_locked),
+            ResetSignal("fast")  .eq(~usb2_locked)
+        ]
+
+        return m
+
+def ULPIResource(name, data_sites, clk_site, dir_site, nxt_site, stp_site, reset_site):
+    """ Generates a set of resources for a ULPI-connected USB PHY. """
+
+    return Resource(name, 0,
+        Subsignal("data",  Pins(data_sites,  dir="io")),
+        Subsignal("clk",   Pins(clk_site,    dir="i" )),
+        Subsignal("dir",   Pins(dir_site,    dir="i" )),
+        Subsignal("nxt",   Pins(nxt_site,    dir="i" )),
+        Subsignal("stp",   Pins(stp_site,    dir="o" )),
+        Subsignal("rst",   PinsN(reset_site, dir="o" )),
+        Attrs(IOSTANDARD="LVCMOS33", SLEW="FAST")
+    )
+
+class NumatoOpsisPlatform(XilinxSpartan6Platform, LUNAPlatform):
+    """ Board description for NumatoOpsis. """
+
+    name                   = "NumatoOpsis"
+
+    device                 = "xc6slx45t"
+    package                = "fgg484"
+    speed                  = "3"
+
+    clock_domain_generator = NumatoOpsisClockDomainGenerator
+    default_usb_connection = "ulpi"
+    ulpi_handle_clocking = False
+
+    resources = [
+
+        # Clock
+        Resource("clk_100MHz", 0, Pins("AB13", dir="i"), Clock(100e6), Attrs(IOSTANDARD="LVCMOS33")),
+
+        # Button
+        *ButtonResources(pins="Y3", attrs=Attrs(IOSTANDARD="LVCMOS15", PULLUP="TRUE")),
+
+        # ULPI phy with input clock operation (ULPI clk generated by link)
+        ULPIResource("ulpi",
+            data_sites="Y14 W14 Y18 W17 AA14 AB14 Y16 W15", clk_site="-",
+            dir_site="W13", nxt_site="V13", stp_site="W18", reset_site="V17",
+        ),
+
+        Resource("forward_ulpi_clk", 0, Pins("Y13", dir="o"), Attrs(IOSTANDARD="LVCMOS33", SLEW="FAST")),
+        Resource("debug_header", 0, Pins("P28_0:SD_DAT1", dir="o"), Attrs(IOSTANDARD="LVCMOS33", SLEW="FAST")),
```

Comment:
For consistency with the other boards (and so things automatically work in some example code), it's probably best to name these `user_io`.